src/main/scala/com/codacy/analysis/cli/Main.scala

```diff
@@ -38,40 +39,65 @@ class MainImpl extends CLIApp {
           "No credentials found.".asLeft[ProjectConfiguration]
         } { _.getRemoteConfiguration }
 
-        val resultsUploader: Either[String, ResultsUploader] = codacyClientOpt.fold {
-          "No credentials found.".asLeft[ResultsUploader]
-        } { codacyClient =>
-          analyse.commitUuid.fold {
-            "No commit option found.".asLeft[ResultsUploader]
-          } { commit =>
-            if (analyse.uploadValue) {
-              new ResultsUploader(commit, codacyClient, None).asRight[String]
-            } else {
-              "Upload option disabled.".asLeft[ResultsUploader]
-            }
-          }
-
-        }
-
-        new AnalyseExecutor(
+        val analysisResults = new AnalyseExecutor(
           analyse.tool,
           analyse.directory,
           formatter,
           analyser,
-          resultsUploader,
           fileCollector,
           remoteProjectConfiguration,
           analyse.parallel).run()
-    }
 
-    Await.result(res, Duration.Inf) match {
-      case Left(e) =>
-        logger.error(s"Failed analysis: $e")
-      case Right(_) =>
-        logger.info("Completed analysis successfully")
+        val uploadResult = uploadResults(codacyClientOpt)(analyse.uploadValue, analyse.commitUuid, analysisResults)
+        Await.result(uploadResult, Duration.Inf) match {
+          case Left(err) if analyse.uploadValue =>
+            logger.error(s"Upload of results failed: $err")
+          case Right(_) if analyse.uploadValue =>
+            logger.info("Completed upload of results to API")
+          case _ =>
+            logger.info("Skipping upload of results to API")
+        }
+
+        System.exit(new Status(analyse.maxAllowedIssues, analyse.failIfIncompleteValue).exitCode(analysisResults))
     }
 
     ()
   }
 
+  private def uploadResults(codacyClientOpt: Option[CodacyClient])(
+    upload: Boolean,
+    commitUuid: Option[String],
+    executorResultsEither: Either[String, Seq[ExecutorResult]]): Future[Either[String, Unit]] = {
+    val resultsUploader: Either[String, ResultsUploader] = codacyClientOpt.fold {
```

Comment:
I would prefer this as a for comprehension with the "Option"s converted to "Either"s using ".toRight", but not a mandatory change.